src/garage/sampler/rl2_sampler.py

```diff
@@ -0,0 +1,224 @@
+
+
+class RL2Sampler:
+    """Sampler which uses VecEnvExecutor to run multiple environments.
+
+    This sampler is for RL^2. See https://arxiv.org/pdf/1611.02779.pdf.
+
+    Args:
+        algo (garage.np.algos.RLAlgorithm): An algorithm instance.
+        env (List[garage.envs.GarageEnv]): Environements to sample from.
+        meta_batch_size (int): Meta batch size for sampling. If it is
+            larger than n_envs, it must be a multiple of n_envs so it can be
+            evenly distributed among environments.
+        n_envs (int): Number of environment instances for sampling. It it is
+            larger than meta_batch_size, it must be a multiple of
+            meta_batch_size so batch can be evenly distributed among
+            environments.
+
+    """
+
+    def __init__(self, algo, env, meta_batch_size, n_envs=None):
+        if n_envs is None:
+            n_envs = singleton_pool.n_parallel * 4
+        self.algo = algo
+        self.envs = env
+
+        self._n_envs = n_envs
+        self._meta_batch_size = meta_batch_size
+        self._vec_env = None
+        self._envs_per_worker = None
+        self._vec_envs_indices = None
+
+    def start_worker(self):
+        """Start workers.
+
+        Raises:
+            ValueError: If meta_batch_size > n_envs and meta_batch_size is not
+                a multiple of n_envs, or if n_envs > meta_batch_size and n_envs
+                is not a multiple of meta_batch_size.
+
+        """
+        if self._meta_batch_size > self._n_envs:
+            if self._meta_batch_size % self._n_envs != 0:
+                raise ValueError(
+                    'meta_batch_size must be a multiple of n_envs')
+            self._envs_per_worker = 1
+            self._vec_envs_indices = np.split(np.arange(self._meta_batch_size),
+                                              self._n_envs)
+        if self._n_envs >= self._meta_batch_size:
+            if self._n_envs % self._meta_batch_size != 0:
+                raise ValueError(
+                    'n_envs must be a multiple of meta_batch_size')
+            self._envs_per_worker = self._n_envs // self._meta_batch_size
+            self._vec_envs_indices = [np.arange(self._meta_batch_size)]
+
+    def shutdown_worker(self):
+        """Shutdown workers."""
+        self._vec_env.close()
+
+    def setup_worker(self, env_indices):
+        """Setup workers.
+
+        Args:
+            env_indices (List[Int]): Indices of environments to be assigned
+                to workers for sampling.
+
+        """
+        vec_envs = []
+        for env_ind in env_indices:
+            vec_env = pickle.loads(pickle.dumps(self.envs[env_ind]))
+            vec_envs.extend([vec_env for env in range(self._envs_per_worker)])
+        seed0 = deterministic.get_seed()
+        if seed0 is not None:
+            for (i, e) in enumerate(vec_envs):
+                e.seed(seed0 + i)
+
+        self._vec_env = VecEnvExecutor(
```

Comment:
Wait, so do you create new environments every worker every time `obtain_samples` is called without ever closing any of them? This seems very likely to lead to significant memory / resource leaks.

Reply:
Good catch